File: c_version/parameter_sweep.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

// Physical constants
#define G 6.67430e-11
#define DT 200.0
#define MOON_MASS (7.342e22 * 50)
#define MOON_RADIUS 1737000.0
#define SPACECRAFT_MASS 1000.0
#define MAX_STEPS 100

// Position and velocity structures
typedef struct {
    double x, y;
} Vector2D;

// Simulation parameters
typedef struct {
    Vector2D spacecraft_pos;
    Vector2D target_center;
    double target_radius;
} SimParams;

// Result structure
typedef struct {
    int target_hit;
    int target_hit_time;
    Vector2D final_pos;
} SimResult;

// Vector operations
Vector2D vector_add(Vector2D a, Vector2D b) {
    Vector2D result = {a.x + b.x, a.y + b.y};
    return result;
}

Vector2D vector_subtract(Vector2D a, Vector2D b) {
    Vector2D result = {a.x - b.x, a.y - b.y};
    return result;
}

Vector2D vector_multiply(Vector2D v, double scalar) {
    Vector2D result = {v.x * scalar, v.y * scalar};
    return result;
}

double vector_magnitude(Vector2D v) {
    return sqrt(v.x * v.x + v.y * v.y);
}

Vector2D vector_normalize(Vector2D v) {
    double mag = vector_magnitude(v);
    if (mag > 0) {
        Vector2D result = {v.x / mag, v.y / mag};
        return result;
    }
    Vector2D zero = {0, 0};
    return zero;
}

// Check if spacecraft hits target
int check_target_hit(Vector2D spacecraft_pos, SimParams params) {
    Vector2D diff = vector_subtract(spacecraft_pos, params.target_center);
    double distance = vector_magnitude(diff);
    return distance <= params.target_radius;
}

// Calculate gravitational force
Vector2D gravitational_force(Vector2D pos1, Vector2D pos2, double mass1, double mass2) {
    Vector2D r_vector = vector_subtract(pos2, pos1);
    double r_magnitude = vector_magnitude(r_vector);
    
    // Prevent collision
    double min_distance = MOON_RADIUS * 1.2;
    if (r_magnitude < min_distance) {
        r_magnitude = min_distance;
    }
    
    double force_magnitude = G * mass1 * mass2 / (r_magnitude * r_magnitude);
    Vector2D force_direction = vector_normalize(r_vector);
    
    return vector_multiply(force_direction, force_magnitude);
}

// Run single simulation
SimResult run_simulation(Vector2D spacecraft_vel, SimParams params) {
    // Initialize positions and velocities
    Vector2D moon_pos = {0, 0};
    Vector2D moon_vel = {1500, 0};  // 1.5 km/s rightward
    Vector2D spacecraft_pos = params.spacecraft_pos;
    
    SimResult result = {0, -1, {0, 0}};
    
    // Run simulation
    for (int step = 0; step < MAX_STEPS; step++) {
        // Calculate gravitational force
        Vector2D force = gravitational_force(spacecraft_pos, moon_pos, 
                                           SPACECRAFT_MASS, MOON_MASS);
        
        // Update spacecraft
        Vector2D acceleration = vector_multiply(force, 1.0 / SPACECRAFT_MASS);
        Vector2D vel_change = vector_multiply(acceleration, DT);
        spacecraft_vel = vector_add(spacecraft_vel, vel_change);
        
        Vector2D pos_change = vector_multiply(spacecraft_vel, DT);
        spacecraft_pos = vector_add(spacecraft_pos, pos_change);
        
        // Update moon (constant velocity)
        Vector2D moon_pos_change = vector_multiply(moon_vel, DT);
        moon_pos = vector_add(moon_pos, moon_pos_change);
        
        // Check target hit
        if (check_target_hit(spacecraft_pos, params)) {
            result.target_hit = 1;
            result.target_hit_time = step;
            result.final_pos = spacecraft_pos;
            break;
        }
    }
    
    result.final_pos = spacecraft_pos;
    return result;
}

int main() {
    printf("🚀 Slingshot Parameter Sweep (C Version)\n");
    printf("=========================================\n");
    
    // Initialize simulation parameters
    SimParams params;
    params.spacecraft_pos.x = 3e7;   // 30M km right
    params.spacecraft_pos.y = -2e7;  // 20M km down
    params.target_center.x = 4e7;    // 40M km right
    params.target_center.y = 3e7;    // 30M km up
    params.target_radius = 1e5;      // 100k km radius
    
    printf("Testing velocity ranges:\n");
    printf("X: -4000 to -1000 m/s (step: 10)\n");
    printf("Y: 1000 to 4000 m/s (step: 10)\n");
    printf("Target: [%.0f, %.0f] km, Radius: %.0f km\n\n", 
           params.target_center.x/1000, params.target_center.y/1000, 
           params.target_radius/1000);
    
    // Open output file
    FILE *output_file = fopen("successful_parameters.txt", "w");
    if (!output_file) {
        printf("Error: Cannot create output file!\n");
        return 1;
    }
    
    // Write header
    fprintf(output_file, "# Successful Slingshot Parameters (C Version)\n");
    fprintf(output_file, "# Format: Vx(m/s), Vy(m/s), Speed(m/s), HitTime(steps)\n");
    fprintf(output_file, "# Target: [4e7, 3e7] km, Radius: 100,000 km\n");
    fprintf(output_file, "# Spacecraft initial position: [3e7, -2e7] km\n\n");
    
    // Parameter ranges
    int vx_min = -4000, vx_max = -1000, vx_step = 10;
    int vy_min = 1000, vy_max = 4000, vy_step = 10;
    
    int total_tests = ((vx_max - vx_min) / vx_step + 1) * ((vy_max - vy_min) / vy_step + 1);
    int successful_count = 0;
    int test_count = 0;
    
    printf("Total parameter combinations: %d\n", total_tests);
    printf("Progress: ");
    fflush(stdout);
    
    clock_t start_time = clock();
    
    // Parameter sweep
    for (int vx = vx_min; vx <= vx_max; vx += vx_step) {
        for (int vy = vy_min; vy <= vy_max; vy += vy_step) {
            test_count++;
            
            // Run simulation
            Vector2D spacecraft_vel = {(double)vx, (double)vy};
            SimResult result = run_simulation(spacecraft_vel, params);
            
            // Check if successful
            if (result.target_hit) {
                successful_count++;
                double speed = sqrt(vx * vx + vy * vy);
                
                // Write to file
                fprintf(output_file, "%6d, %6d, %7.0f, %3d\n", 
                       vx, vy, speed, result.target_hit_time);
            }
            
            // Progress reporting (every 5%)
            if (test_count % (total_tests / 20) == 0 || test_count == total_tests) {
                printf(".");
                fflush(stdout);
            }
        }
    }
    
    clock_t end_time = clock();
    double execution_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
    
    fclose(output_file);
    
    printf("\n\n✅ Parameter sweep completed!\n");
    printf("Total tests: %d\n", total_tests);
    printf("Successful parameters: %d\n", successful_count);
    printf("Success rate: %.2f%%\n", (double)successful_count / total_tests * 100);
    printf("Execution time: %.2f seconds\n", execution_time);
    printf("💾 Results saved to 'successful_parameters.txt'\n");
    
    // Show first few successful parameters if any
    if (successful_count > 0) {
        printf("\n📊 First few successful parameters:\n");
        printf("Vx (m/s) | Vy (m/s) | Speed (m/s) | Hit Time\n");
        printf("---------|----------|-------------|----------\n");
        
        // Re-read and display first 5 results
        FILE *read_file = fopen("successful_parameters.txt", "r");
        if (read_file) {
            char line[256];
            int count = 0;
            
            // Skip header lines
            while (fgets(line, sizeof(line), read_file) && line[0] == '#') {}
            
            // Read and display first 5 data lines
            do {
                if (line[0] != '#' && line[0] != '\n') {
                    int vx, vy, hit_time;
                    double speed;
                    if (sscanf(line, "%d, %d, %lf, %d", &vx, &vy, &speed, &hit_time) == 4) {
                        printf("%8d | %8d | %11.0f | %8d\n", vx, vy, speed, hit_time);
                        count++;
                    }
                }
            } while (fgets(line, sizeof(line), read_file) && count < 5);
            
            if (successful_count > 5) {
                printf("... and %d more\n", successful_count - 5);
            }
            
            fclose(read_file);
        }
    }
    
    printf("\n🎯 Analysis complete!\n");
    return 0;
} 

File: parameter_sweep.py
# -*- coding: utf-8 -*-
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Circle
import math
import os
from PIL import Image

class SlingshotParameterSweep:
    def __init__(self):
        # Enhanced physics constants for dramatic effect
        self.G = 6.67430e-11
        self.dt = 200  # Time step (seconds)
        
        # Enhanced mode - dramatic effect with clear U-turn
        self.moon_mass = 7.342e22 * 50  # 50x real moon mass for dramatic effect
        self.moon_radius = 1737000  # Real moon radius
        
        # Fixed parameters (same for all tests)
        self.spacecraft_mass = 1000
        self.spacecraft_pos = np.array([3e7, -2e7], dtype=float)  # 30M km ahead, 20M km below
        
        # Target zone in the upper right area
        self.target_center = np.array([4e7, 3e7], dtype=float)  # 40M km right, 30M km up
        self.target_radius = 1e5  # 100k km radius - small target zone
        
        # Results storage
        self.successful_params = []
        self.all_results = []
        
    def calculate_escape_velocity(self, distance):
        """Calculate escape velocity at given distance"""
        return np.sqrt(2 * self.G * self.moon_mass / distance)
    
    def check_target_hit(self, spacecraft_pos):
        """Check if spacecraft has hit the target"""
        distance_to_target = np.linalg.norm(spacecraft_pos - self.target_center)
        return distance_to_target <= self.target_radius
    
    def gravitational_force(self, pos1, pos2, mass1, mass2):
        """Calculate gravitational force"""
        r_vector = pos2 - pos1
        r_magnitude = np.linalg.norm(r_vector)
        
        # Prevent collision
        min_distance = self.moon_radius * 1.2
        if r_magnitude < min_distance:
            r_magnitude = min_distance
            
        force_magnitude = self.G * mass1 * mass2 / (r_magnitude ** 2)
        force_direction = r_vector / r_magnitude
        
        return force_magnitude * force_direction
    
    def run_single_simulation(self, spacecraft_vel, max_steps=100):
        """Run a single simulation with given initial velocity"""
        # Initialize positions and velocities
        moon_pos = np.array([0, 0], dtype=float)
        moon_vel = np.array([1500, 0], dtype=float)  # 1.5 km/s rightward
        spacecraft_pos = self.spacecraft_pos.copy()
        spacecraft_vel = np.array(spacecraft_vel, dtype=float)
        
        target_hit = False
        target_hit_time = None
        
        # Run simulation
        for step in range(max_steps):
            # Calculate gravitational force
            force = self.gravitational_force(
                spacecraft_pos, moon_pos,
                self.spacecraft_mass, self.moon_mass
            )
            
            # Update spacecraft
            acceleration = force / self.spacecraft_mass
            spacecraft_vel += acceleration * self.dt
            spacecraft_pos += spacecraft_vel * self.dt
            
            # Update moon (constant velocity)
            moon_pos += moon_vel * self.dt
            
            # Check target hit
            if self.check_target_hit(spacecraft_pos):
                target_hit = True
                target_hit_time = step
                break
        
        return {
            'target_hit': target_hit,
            'target_hit_time': target_hit_time,
            'final_pos': spacecraft_pos.copy(),
            'final_distance_to_target': np.linalg.norm(spacecraft_pos - self.target_center)
        }
    
    def parameter_sweep(self):
        """Sweep through parameter space"""
        print("🔍 Starting parameter sweep...")
        print("Testing spacecraft velocity ranges:")
        print("X: -4000 to -1000 m/s (step: 10)")
        print("Y: 1000 to 4000 m/s (step: 10)")
        print()
        
        # Define parameter ranges
        x_velocities = np.arange(-4000, -1000 + 10, 10)  # -4000 to -1000, step 10
        y_velocities = np.arange(1000, 4000 + 10, 10)    # 1000 to 4000, step 10
        
        total_tests = len(x_velocities) * len(y_velocities)
        print(f"Total parameter combinations to test: {total_tests}")
        print(f"X velocities: {len(x_velocities)} points")
        print(f"Y velocities: {len(y_velocities)} points")
        print()
        
        successful_count = 0
        test_count = 0
        
        # Progress tracking
        progress_interval = max(1, total_tests // 20)  # Show progress every 5%
        
        for i, vx in enumerate(x_velocities):
            for j, vy in enumerate(y_velocities):
                test_count += 1
                
                # Run simulation
                result = self.run_single_simulation([vx, vy])
                
                # Store result
                result_data = {
                    'vx': vx,
                    'vy': vy,
                    'speed': np.sqrt(vx**2 + vy**2),
                    'target_hit': result['target_hit'],
                    'target_hit_time': result['target_hit_time'],
                    'final_distance': result['final_distance_to_target'] / 1000  # km
                }
                
                self.all_results.append(result_data)
                
                if result['target_hit']:
                    successful_count += 1
                    self.successful_params.append(result_data)
                
                # Progress reporting
                if test_count % progress_interval == 0 or test_count == total_tests:
                    progress = (test_count / total_tests) * 100
                    print(f"Progress: {progress:.1f}% ({test_count}/{total_tests}) - "
                          f"Successful: {successful_count}")
        
        print(f"\n✅ Parameter sweep completed!")
        print(f"Total tests: {total_tests}")
        print(f"Successful parameters: {successful_count}")
        print(f"Success rate: {(successful_count/total_tests)*100:.2f}%")
        
        return self.successful_params, self.all_results
    
    def analyze_results(self):
        """Analyze the sweep results"""
        if not self.successful_params:
            print("❌ No successful parameters found!")
            return
        
        print(f"\n📊 Analysis of {len(self.successful_params)} successful parameters:")
        print("=" * 60)
        
        # Convert to arrays for analysis
        vx_success = [p['vx'] for p in self.successful_params]
        vy_success = [p['vy'] for p in self.successful_params]
        speeds_success = [p['speed'] for p in self.successful_params]
        hit_times = [p['target_hit_time'] for p in self.successful_params]
        
        print(f"X velocity range: {min(vx_success)} to {max(vx_success)} m/s")
        print(f"Y velocity range: {min(vy_success)} to {max(vy_success)} m/s")
        print(f"Speed range: {min(speeds_success):.0f} to {max(speeds_success):.0f} m/s")
        print(f"Hit time range: {min(hit_times)} to {max(hit_times)} steps")
        print()
        
        # Show first 10 successful parameters
        print("🎯 First 10 successful parameters:")
        print("Vx (m/s) | Vy (m/s) | Speed (m/s) | Hit Time (steps)")
        print("-" * 55)
        for i, param in enumerate(self.successful_params[:10]):
            print(f"{param['vx']:8.0f} | {param['vy']:8.0f} | {param['speed']:11.0f} | {param['target_hit_time']:13}")
        
        if len(self.successful_params) > 10:
            print(f"... and {len(self.successful_params) - 10} more")
        
        return vx_success, vy_success, speeds_success
    
    def create_visualization(self):
        """Create visualization of parameter space"""
        if not self.all_results:
            print("No results to visualize!")
            return
        
        print("\n🎨 Creating parameter space visualization...")
        
        # Extract data
        vx_all = [r['vx'] for r in self.all_results]
        vy_all = [r['vy'] for r in self.all_results]
        success_all = [r['target_hit'] for r in self.all_results]
        
        # Create figure
        plt.style.use('dark_background')
        fig, ax = plt.subplots(1, 1, figsize=(12, 10))
        fig.patch.set_facecolor('black')
        ax.set_facecolor('black')
        
        # Plot all points
        successful_vx = [vx_all[i] for i in range(len(vx_all)) if success_all[i]]
        successful_vy = [vy_all[i] for i in range(len(vy_all)) if success_all[i]]
        failed_vx = [vx_all[i] for i in range(len(vx_all)) if not success_all[i]]
        failed_vy = [vy_all[i] for i in range(len(vy_all)) if not success_all[i]]
        
        # Plot failed points (red)
        ax.scatter(failed_vx, failed_vy, c='red', s=1, alpha=0.3, label='Failed')
        
        # Plot successful points (green)
        ax.scatter(successful_vx, successful_vy, c='lime', s=3, alpha=0.8, label='Success')
        
        # Mark known successful cases
        ax.scatter([-2800], [1600], c='yellow', s=100, marker='*', 
                  label='Known Success', edgecolors='black', linewidth=1)
        ax.scatter([-2600], [1900], c='orange', s=100, marker='x', 
                  label='Known Fail', linewidth=3)
        
        ax.set_xlabel('X Velocity (m/s)', fontsize=14, color='white')
        ax.set_ylabel('Y Velocity (m/s)', fontsize=14, color='white')
        ax.set_title('Slingshot Parameter Space - Success/Fail Map', 
                    fontsize=16, color='white', pad=20)
        ax.grid(True, alpha=0.3)
        ax.legend(fontsize=12)
        
        # Add statistics text
        success_rate = (len(successful_vx) / len(vx_all)) * 100
        stats_text = f'Total tests: {len(vx_all)}\nSuccessful: {len(successful_vx)}\nSuccess rate: {success_rate:.1f}%'
        ax.text(0.02, 0.98, stats_text, transform=ax.transAxes, 
                fontsize=12, color='white', verticalalignment='top',
                bbox=dict(boxstyle="round,pad=0.5", facecolor='black', alpha=0.8))
        
        plt.tight_layout()
        
        # Save the plot before showing (to avoid potential issues)
        try:
            plt.savefig('parameter_sweep_results.png', dpi=300, 
                       facecolor='black', edgecolor='none', 
                       bbox_inches='tight', pad_inches=0.1)
            print("📁 Visualization saved as 'parameter_sweep_results.png'")
        except Exception as e:
            print(f"Warning: Could not save plot: {e}")
        
        # Also save as PDF for better quality
        try:
            plt.savefig('parameter_sweep_results.pdf', 
                       facecolor='black', edgecolor='none', 
                       bbox_inches='tight', pad_inches=0.1)
            print("📁 High-quality version saved as 'parameter_sweep_results.pdf'")
        except Exception as e:
            print(f"Warning: Could not save PDF: {e}")
        
        # Save a white background version as backup
        try:
            # Temporarily change colors for white background version
            fig.patch.set_facecolor('white')
            ax.set_facecolor('white')
            ax.set_title('Slingshot Parameter Space - Success/Fail Map', 
                        fontsize=16, color='black', pad=20)
            ax.set_xlabel('X Velocity (m/s)', fontsize=14, color='black')
            ax.set_ylabel('Y Velocity (m/s)', fontsize=14, color='black')
            ax.tick_params(colors='black')
            
            # Update stats text color
            for text in ax.texts:
                if 'Total tests:' in text.get_text():
                    text.set_color('black')
                    text.get_bbox_patch().set_facecolor('white')
                    text.get_bbox_patch().set_edgecolor('black')
            
            plt.savefig('parameter_sweep_results_white.png', dpi=300, 
                       facecolor='white', edgecolor='none', 
                       bbox_inches='tight', pad_inches=0.1)
            print("📁 White background version saved as 'parameter_sweep_results_white.png'")
            
            # Restore original colors
            fig.patch.set_facecolor('black')
            ax.set_facecolor('black')
            ax.set_title('Slingshot Parameter Space - Success/Fail Map', 
                        fontsize=16, color='white', pad=20)
            ax.set_xlabel('X Velocity (m/s)', fontsize=14, color='white')
            ax.set_ylabel('Y Velocity (m/s)', fontsize=14, color='white')
            ax.tick_params(colors='white')
            
        except Exception as e:
            print(f"Warning: Could not save white background version: {e}")
        
        plt.show()
    
    def save_results(self, filename='successful_parameters.txt'):
        """Save successful parameters to file"""
        if not self.successful_params:
            print("No successful parameters to save!")
            return
        
        with open(filename, 'w') as f:
            f.write("# Successful Slingshot Parameters\n")
            f.write("# Format: Vx(m/s), Vy(m/s), Speed(m/s), HitTime(steps)\n")
            f.write("# Target: [4e7, 3e7] km, Radius: 100,000 km\n")
            f.write("# Spacecraft initial position: [3e7, -2e7] km\n\n")
            
            for param in self.successful_params:
                f.write(f"{param['vx']:6.0f}, {param['vy']:6.0f}, {param['speed']:7.0f}, {param['target_hit_time']:3}\n")
        
        print(f"💾 Successful parameters saved to '{filename}'")

def main():
    print("🚀 Slingshot Parameter Sweep Analysis")
    print("=" * 50)
    
    # Initialize sweep
    sweep = SlingshotParameterSweep()
    
    # Run parameter sweep
    successful_params, all_results = sweep.parameter_sweep()
    
    # Analyze results
    if successful_params:
        sweep.analyze_results()
        
        # Create visualization
        sweep.create_visualization()
        
        # Save results
        sweep.save_results()
    else:
        print("\n❌ No successful parameters found in the tested range!")
        print("Consider:")
        print("- Expanding the search range")
        print("- Reducing the step size")
        print("- Increasing the target radius")
        print("- Increasing the simulation steps")
    
    print("\n🎯 Parameter sweep analysis complete!")

if __name__ == "__main__":
    main() 